### PR TITLE
Navigation Bar + Login Redirect

### LINKED ISSUE #32: Navigation: Navigation Component + Log Out
# Overview

### `NavBar`

Create a `NavBar` component and implement the navigation bar design ([Figma](https://www.figma.com/file/06vEDN5gtlHlC5hxNXMb4g/USHS-Designs?node-id=337%3A10458&mode=dev)). Use this new component in the `Page` component, which should be used on all pages that need a navbar

### Log Out Functionality

Implement the Log Out popup ([Figma](https://www.figma.com/file/06vEDN5gtlHlC5hxNXMb4g/USHS-Designs?node-id=341%3A5902&mode=dev)).

When "Log out" is clicked, call [`signOut`](https://firebase.google.com/docs/auth/web/google-signin#next_steps) from Firebase and redirect the user to `/login`.

### Protected Pages

If a signed-out user manually navigates to protected pages (e.g. `/` or `/profile`), they should be redirected to `/login`. You can use the [`Navigate`](https://reactrouter.com/en/main/components/navigate) component form react-router ([example](https://github.com/TritonSE/TAP-Foster-Dog-Matching/blob/101fb04c11dd45d90e6956a7c417414006bbafe4/client/src/App.js#L28C40-L28C40)).

## Backend

N/A

## Acceptance Criteria
- [x] Navigation bar implemented
- [x] Log out works as expected
- [x] If the user is not signed in, they should not be able to access protected pages

### PR DESCRIPTION
**Pull request by Azhou2023**
## Tracking Info

Resolves #32  

## Changes

Added navigation bar for all pages and a rudimentary profile page. Changed routes so that if a user isn't logged in, they are immediately redirected to the login page no matter what path they try to visit. 

## Testing

Made sure that if there was no current user, all manual paths (valid or invalid) would redirect to the login page. Also ensured that if a user was logged in, they could navigate to the login page and back with no barriers. For the nav bar, I tested that the paths were correct and that the log out popup stopped any interaction with the rest of the screen (and that logging out actually works and immediately redirects to the login page).

## Confirmation of Change

Screenshot_18
Screenshot_19
Screenshot_20

### CODE REVIEW

File: frontend/src/contexts/DataContext.tsx
Comment (Azhou2023) [inline frontend/src/contexts/DataContext.tsx:1]: Ignore this file, not really part of this pr (from the frontend contexts branch). Just commented out for the pr, will revert as soon as the other branch is merged.

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -1,22 +1,49 @@
...
+          <>
+            <Route path="*" element={<Navigate replace to="/login" />} />
+            <Route path="/login" element={<Login />} />
+          </>
+        )}
+
+        {signedIn && (
+          <>
             <Route path="/" element={<Home />} />
             <Route path="/login" element={<Login />} />
```
Comment (petabite) [inline frontend/src/App.tsx]: What do you think about not letting the user go to `/login` if they are already signed in?

File: frontend/src/pages/Login.tsx [hunk 2]
```diff
@@ -24,9 +24,8 @@ const Wrapper = styled.div`
   align-items: center;
   gap: 13.67vh;
   padding: 20vh 364.6px 0vh 365px;
-  // max-height: 70vh;
 `;
-const Button = styled.button`
+export const Button = styled.button`
```
Comment (petabite) [inline frontend/src/pages/Login.tsx]: Can we extract this to the `components` folder? We are probably going to use this a lot throughout the app

**Review by petabite — CHANGES_REQUESTED** [review]
Nice! Just a couple things:

File: frontend/src/App.tsx (hunk 1 above)
Comment (Azhou2023) [reply frontend/src/App.tsx]: Sounds good; I just made it so that any invalid route when logged in redirects to the home page, is that fine?

File: frontend/src/pages/Login.tsx (hunk 2 above)
Comment (Azhou2023) [reply frontend/src/pages/Login.tsx]: Also sounds good; for now, I didn't make Button.tsx a full fledged component class, just a file with a few exported styled components; is that alright? I couldn't quite figure out how to make it a component that accepts props that can also be further modifed by styled components in different classes (because from what I've seen on Figma not all the button designs are consistent).

File: frontend/src/components/NavBar.tsx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        console.log(error);
+      });
+  };
+
+  return (
+    <div>
+      <NavbarItems>
+        <LeftWrapper>
+          <Icon src="USHSLogo2.png" />
+          {page === "Home" ? (
```
Comment (nbpillai) [inline frontend/src/components/NavBar.tsx]: I think here this could maybe be simplified by taking out the ternary and `Highlighted` component. Instead you could add an `active` prop to style your `Link` component. Then the active prop could be passed in according to what page we are on.  
```
const Link = styled.a<{ active: boolean }>`
  text-decoration: none;

  ${(props) =>
    props.active
      ? css`
          color: #b64201;
        `
      : css`
          color: black;
        `}
`;
```

```
<Link href="/" active={page === "Home"}> Home </Link>
<Link href="/profile" active={page === "Profile"}> Profile </Link>
```

**Review by nbpillai — COMMENTED** [review]
everything looks really good! I just added one suggestion for simplifying the nav bar tab color change when you switch from the home to profile page.

File: frontend/src/App.tsx (hunk 1 above)
Comment (petabite) [reply frontend/src/App.tsx]: Yep, that's perfect.

File: frontend/src/pages/Login.tsx (hunk 2 above)
Comment (petabite) [reply frontend/src/pages/Login.tsx]: That's fine. If we find we need to consolidate into one component, we can refactor down the line.

File: frontend/src/components/NavBar.tsx (hunk 3 above)
Comment (Azhou2023) [reply frontend/src/components/NavBar.tsx]: Thanks so much for the suggestion, that's a lot cleaner

File: frontend/src/components/Button.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  background-color: ${(props) => (props.kind === "primary" ? "#b64201" : "#ffffff")};
+  color: ${(props) => (props.kind === "primary" ? "#ffffff" : "#b64201")};
+  border: 1px solid #b64201;
+  border-radius: 14px;
+  font-weight: 500;
+  font-size: 18px;
+  letter-spacing: 0.32px;
+  cursor: pointer;
+  white-space: nowrap;
+`;
```
Comment (petabite) [inline frontend/src/components/Button.tsx:14]: 🎉

**Review by petabite — APPROVED** [review]
LGTM 🚢

**Review by nbpillai — APPROVED** [review]
Lgtm!
